File: dsa/li18_string/L659_EncodeAndDecodeStrings.java
package li18_string;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class L659_EncodeAndDecodeStrings {

	public static void main(String[] args) {
		L659_EncodeAndDecodeStrings obj = new L659_EncodeAndDecodeStrings();
		String encoded = obj.encode(new ArrayList<String>(Arrays.asList("lintsfsdkjfsdkfjsdkfsd","code","love","you")));
		System.out.println(encoded);
		System.out.println(obj.decode(encoded));
	}

    // Problem: https://www.lintcode.com/problem/659/
    // Idea: You can't use simple delimieter as strings can also contain that delimiter.
    // You can solve that, you have to store no of chars in each word.
    // But this number can agian conflict with ints that can be there in words like ab23d.
    // Hence, introduce a simbal after number to idetify that this number is used by us.
	public String encode(List<String> strs) {
        StringBuffer result = new StringBuffer();
        String delimeter = "#";
        for(int i=0; i<strs.size(); i++){
            result.append(strs.get(i).length());
            result.append(delimeter);
            result.append(strs.get(i));
        }
        return result.toString();
    }

    public List<String> decode(String str) {
        List<String> result = new ArrayList<>();
        int i=0;
        while(i<str.length()){
            int delInd = str.indexOf('#', i);
            int length = Integer.parseInt(str.substring(i, delInd));

            int begInd = delInd+1;
            int endInd = begInd+length;
            result.add(str.substring(begInd, endInd));
            i = endInd;
        }
        return result;
    }
}

File: dsa/concepts/18_strings/python/l659_encode_and_decode_strings.py
# Problem: https://www.lintcode.com/problem/659/
# Idea: You can't use simple delimieter as strings can also contain that delimiter.
# You can solve that, you have to store no of chars in each word.
# But this number can agian conflict with ints that can be there in words like ab23d.
# Hence, introduce a simbal after number to idetify that this number is used by us.
def encode(strs):
    result = []
    delimiter = "#"
    for s in strs:
        result.append(str(len(s)))
        result.append(delimiter)
        result.append(s)
    return "".join(result)


def decode(s):
    result = []
    i = 0
    n = len(s)

    while i < n:
        del_ind = s.index('#', i)
        length = int(s[i:del_ind])

        beg_ind = del_ind + 1
        end_ind = beg_ind + length
        result.append(s[beg_ind:end_ind])

        i = end_ind

    return result


# ------------------ Test (Same as Java main) ------------------

encoded = encode(["lintsfsdkjfsdkfjsdkfsd", "code", "love", "you"])
print(encoded)
print(decode(encoded))
